File: app/src/lib/volleyball-stats.ts
import type {
  VolleyballFormattedStats,
  VolleyballTeamStats,
  VolleyballFormMatch,
  VolleyballHeadToHead,
  VolleyballSetBreakdown,
  VolleyballTopScorer,
  VolleyballProgressionEntry,
  VolleyballRecordResult,
  StreakInfo,
  CalendarEvent,
} from '@/types/events';
import { eventsData } from '@/data/events';
import { MONTH_ORDER } from '@/data/month-config';

type VolleyballSport = 'volleyball-men' | 'volleyball-women';

export function parseVolleyballScore(score: string, location: string): [number, number] | null {
  if (!score || !score.includes('-')) return null;
  const parts = score.split('-').map(s => parseInt(s.trim()));
  if (parts.length !== 2 || isNaN(parts[0]) || isNaN(parts[1])) return null;
  return location === 'home' ? [parts[0], parts[1]] : [parts[1], parts[0]];
}

export function getVolleyballResult(setsFor: number, setsAgainst: number): 'W' | 'L' {
  return setsFor > setsAgainst ? 'W' : 'L';
}

export function calculateSetBreakdown(
  matches: CalendarEvent[],
  location?: string
): VolleyballSetBreakdown {
  const breakdown: VolleyballSetBreakdown = {
    threeZero: 0, threeOne: 0, threeTwo: 0,
    zeroThree: 0, oneThree: 0, twoThree: 0,
  };

  for (const match of matches) {
    if (location && match.location !== location) continue;
    if (!match.score) continue;

    const parsed = parseVolleyballScore(match.score, match.location);
    if (!parsed) continue;
    const [setsFor, setsAgainst] = parsed;

    const key = `${setsFor}-${setsAgainst}`;
    switch (key) {
      case '3-0': breakdown.threeZero++; break;
      case '3-1': breakdown.threeOne++; break;
      case '3-2': breakdown.threeTwo++; break;
      case '0-3': breakdown.zeroThree++; break;
      case '1-3': breakdown.oneThree++; break;
      case '2-3': breakdown.twoThree++; break;
    }
  }

  return breakdown;
}

export function aggregateTopScorers(matches: CalendarEvent[]): VolleyballTopScorer[] {
  const scorerMap: Record<string, { totalPoints: number; matchesPlayed: number }> = {};

  for (const match of matches) {
    if (!match.vbScorers) continue;

    // "Our" team: home when location is home, away when location is away
    const ourTeam = match.location;
    const matchScorers = new Set<string>();

    for (const scorer of match.vbScorers) {
      if (scorer.team !== ourTeam) continue;

      if (!scorerMap[scorer.name]) {
        scorerMap[scorer.name] = { totalPoints: 0, matchesPlayed: 0 };
      }
      scorerMap[scorer.name].totalPoints += scorer.points;
      matchScorers.add(scorer.name);
    }

    // Increment matchesPlayed for each unique player in this match
    for (const name of matchScorers) {
      scorerMap[name].matchesPlayed++;
    }
  }

  return Object.entries(scorerMap)
    .map(([name, data]) => ({ name, ...data }))
    .sort((a, b) => b.totalPoints - a.totalPoints)
    .slice(0, 10);
}

function createEmptyTeamStats(): VolleyballTeamStats {
  return {
    played: 0, wins: 0, losses: 0,
    setsWon: 0, setsLost: 0,
    setWinPercentage: 0, winPercentage: 0,
    pointsScored: 0, pointsConceded: 0,
  };
}

function formatSetScores(sets: { home: number; away: number }[], location: string): string {
  return sets.map(s =>
    location === 'home' ? `${s.home}-${s.away}` : `${s.away}-${s.home}`
  ).join(', ');
}

export function calculateVolleyballStatistics(sport: VolleyballSport): VolleyballFormattedStats {
  const overall = createEmptyTeamStats();
  const home = createEmptyTeamStats();
  const away = createEmptyTeamStats();

  interface Match {
    month: string;
    day: number;
    opponent: string;
    location: 'home' | 'away';
    score: string;
    sets?: { home: number; away: number }[];
    vbScorers?: { name: string; points: number; team: 'home' | 'away' }[];
  }

  const matches: Match[] = [];

  for (const monthName of MONTH_ORDER) {
    const events = eventsData[monthName] || [];
    for (const ev of events) {
      if (ev.sport === sport && ev.status === 'played' && ev.score) {
        matches.push({
          month: monthName,
          day: ev.day,
          opponent: ev.opponent,
          location: ev.location,
          score: ev.score,
          sets: ev.sets,
          vbScorers: ev.vbScorers,
        });
      }
    }
  }

  const allForm: VolleyballFormMatch[] = [];
  const h2hMap: Record<string, VolleyballHeadToHead> = {};
  const seasonProgress: VolleyballProgressionEntry[] = [];
  let cumulativeSetsWon = 0;

  // Streak tracking
  let currentWinStreak = 0;
  let currentLossStreak = 0;
  let longestWinStreak = 0;

  // Records tracking
  let biggestWin: (VolleyballRecordResult & { margin: number; rallyDiff: number }) | null = null;
  let heaviestDefeat: (VolleyballRecordResult & { margin: number; rallyDiff: number }) | null = null;

  // CalendarEvent array for set breakdown and top scorers
  const calendarEvents: CalendarEvent[] = [];

  for (const match of matches) {
    const parsed = parseVolleyballScore(match.score, match.location);
    if (!parsed) continue;
    const [setsFor, setsAgainst] = parsed;

    const result = getVolleyballResult(setsFor, setsAgainst);

    // Rally points from sets
    let rallyFor = 0;
    let rallyAgainst = 0;
    if (match.sets) {
      for (const set of match.sets) {
        if (match.location === 'home') {
          rallyFor += set.home;
          rallyAgainst += set.away;
        } else {
          rallyFor += set.away;
          rallyAgainst += set.home;
        }
      }
    }

    // Overall stats
    overall.played++;
    if (result === 'W') overall.wins++;
    else overall.losses++;
    overall.setsWon += setsFor;
    overall.setsLost += setsAgainst;
    overall.pointsScored += rallyFor;
    overall.pointsConceded += rallyAgainst;

    // Home/away stats
    const loc = match.location === 'home' ? home : away;
    loc.played++;
    if (result === 'W') loc.wins++;
    else loc.losses++;
    loc.setsWon += setsFor;
    loc.setsLost += setsAgainst;
    loc.pointsScored += rallyFor;
    loc.pointsConceded += rallyAgainst;

    // Form
    allForm.push({
      result,
      opponent: match.opponent,
      score: match.score,
      location: match.location,
      month: match.month,
      day: match.day,
    });

    // Head-to-head
    if (!h2hMap[match.opponent]) {
      h2hMap[match.opponent] = {
        opponent: match.opponent, played: 0, wins: 0, losses: 0,
        setsWon: 0, setsLost: 0,
      };
    }
    const h2h = h2hMap[match.opponent];
    h2h.played++;
    if (result === 'W') h2h.wins++;
    else h2h.losses++;
    h2h.setsWon += setsFor;
    h2h.setsLost += setsAgainst;

    // Streaks
    if (result === 'W') {
      currentWinStreak++;
      currentLossStreak = 0;
      if (currentWinStreak > longestWinStreak) longestWinStreak = currentWinStreak;
    } else {
      currentLossStreak++;
      currentWinStreak = 0;
    }

    // Records
    const setMargin = setsFor - setsAgainst;
    const rallyDiff = rallyFor - rallyAgainst;
    const setScores = match.sets ? formatSetScores(match.sets, match.location) : '';

    if (setMargin > 0) {
      if (!biggestWin || setMargin > biggestWin.margin ||
          (setMargin === biggestWin.margin && rallyDiff > biggestWin.rallyDiff)) {
        biggestWin = { opponent: match.opponent, score: match.score, setScores, margin: setMargin, rallyDiff };
      }
    }
    if (setMargin < 0) {
      const absMargin = Math.abs(setMargin);
      if (!heaviestDefeat || absMargin > heaviestDefeat.margin ||
          (absMargin === heaviestDefeat.margin && rallyDiff < -heaviestDefeat.rallyDiff)) {
        heaviestDefeat = { opponent: match.opponent, score: match.score, setScores, margin: absMargin, rallyDiff: Math.abs(rallyDiff) };
      }
    }

    // Season progress
    cumulativeSetsWon += setsFor;
    seasonProgress.push({ match: overall.played, setsWon: cumulativeSetsWon, opponent: match.opponent });

    // Build CalendarEvent for set breakdown / top scorers
    calendarEvents.push({
      title: '', subtitle: '', opponent: match.opponent,
      sport, day: match.day, isMeeting: false,
      location: match.location,
      score: match.score,
      sets: match.sets,
      vbScorers: match.vbScorers,
    });
  }

  // Percentages
  const pct = (n: number, d: number) => d === 0 ? 0 : Math.round((n / d) * 100);
  overall.winPercentage = pct(overall.wins, overall.played);
  overall.setWinPercentage = pct(overall.setsWon, overall.setsWon + overall.setsLost);
  home.winPercentage = pct(home.wins, home.played);
  home.setWinPercentage = pct(home.setsWon, home.setsWon + home.setsLost);
  away.winPercentage = pct(away.wins, away.played);
  away.setWinPercentage = pct(away.setsWon, away.setsWon + away.setsLost);

  // Current streak
  const currentStreak: StreakInfo = currentWinStreak > 0
    ? { type: 'W', count: currentWinStreak }
    : { type: 'L', count: currentLossStreak };

  // Clean up records (remove internal margin/rallyDiff)
  const cleanRecord = (r: (VolleyballRecordResult & { margin: number; rallyDiff: number }) | null): VolleyballRecordResult | null => {
    if (!r) return null;
    return { opponent: r.opponent, score: r.score, setScores: r.setScores };
  };

  return {
    overall,
    home,
    away,
    setBreakdown: calculateSetBreakdown(calendarEvents),
    recentForm: allForm.slice(-5).reverse(),
    headToHead: Object.values(h2hMap).sort((a, b) => b.played - a.played),
    streaks: { currentStreak, longestWinStreak },
    records: { biggestWin: cleanRecord(biggestWin), heaviestDefeat: cleanRecord(heaviestDefeat) },
    topScorers: aggregateTopScorers(calendarEvents),
    seasonProgress,
  };
}

export function getNextVolleyballMatch(sport: VolleyballSport): {
  opponent: string;
  date: Date;
  location: 'home' | 'away';
  venue?: string;
} | null {
  for (const monthName of MONTH_ORDER) {
    const events = eventsData[monthName] || [];
    for (const ev of events) {
      if (ev.sport === sport && ev.status === 'upcoming') {
        // Get month info for date construction
        const monthMap: Record<string, number> = {
          september: 8, october: 9, november: 10, december: 11,
          january: 0, february: 1, march: 2, april: 3,
          may: 4, june: 5, july: 6, august: 7,
        };
        const monthIndex = monthMap[monthName];
        // Season: sep-dec = start year, jan-aug = end year
        const year = monthIndex >= 8 ? new Date().getFullYear() : new Date().getFullYear() + 1;

        return {
          opponent: ev.opponent,
          date: new Date(year, monthIndex, ev.day),
          location: ev.location,
          venue: ev.venue,
        };
      }
    }
  }
  return null;
}

File: flutter/lib/logic/volleyball_stats.dart
/// Port of app/src/lib/volleyball-stats.ts.
library;

import '../data/constants.dart';
import '../data/events_repository.dart';
import '../models/events.dart';
import 'football_stats.dart' show StreakInfo, parseScore;

class VolleyballTeamStats {
  int played = 0, wins = 0, losses = 0;
  int setsWon = 0, setsLost = 0;
  int pointsScored = 0, pointsConceded = 0;
  int get winPercentage => played == 0 ? 0 : (wins / played * 100).round();
  int get setWinPercentage {
    final total = setsWon + setsLost;
    return total == 0 ? 0 : (setsWon / total * 100).round();
  }
}

class VolleyballSetBreakdown {
  int threeZero = 0, threeOne = 0, threeTwo = 0;
  int zeroThree = 0, oneThree = 0, twoThree = 0;
}

class VolleyballFormMatch {
  const VolleyballFormMatch({required this.result, required this.opponent, required this.score, required this.location, required this.month, required this.day});
  final String result; // 'W' | 'L'
  final String opponent;
  final String score;
  final MatchLocation location;
  final String month;
  final int day;
}

class VolleyballHeadToHead {
  VolleyballHeadToHead(this.opponent);
  final String opponent;
  int played = 0, wins = 0, losses = 0, setsWon = 0, setsLost = 0;
}

class VolleyballRecordResult {
  const VolleyballRecordResult({required this.opponent, required this.score, required this.setScores});
  final String opponent;
  final String score;
  final String setScores;
}

class VolleyballTopScorer {
  const VolleyballTopScorer({required this.name, required this.totalPoints, required this.matchesPlayed});
  final String name;
  final int totalPoints;
  final int matchesPlayed;
}

class VolleyballStats {
  VolleyballStats({
    required this.overall,
    required this.home,
    required this.away,
    required this.setBreakdown,
    required this.recentForm,
    required this.headToHead,
    required this.currentStreak,
    required this.longestWinStreak,
    required this.biggestWin,
    required this.heaviestDefeat,
    required this.topScorers,
  });

  final VolleyballTeamStats overall;
  final VolleyballTeamStats home;
  final VolleyballTeamStats away;
  final VolleyballSetBreakdown setBreakdown;
  final List<VolleyballFormMatch> recentForm;
  final List<VolleyballHeadToHead> headToHead;
  final StreakInfo currentStreak;
  final int longestWinStreak;
  final VolleyballRecordResult? biggestWin;
  final VolleyballRecordResult? heaviestDefeat;
  final List<VolleyballTopScorer> topScorers;
}

String _formatSetScores(List<VolleyballSet> sets, MatchLocation location) =>
    sets.map((s) => location == MatchLocation.home ? '${s.home}-${s.away}' : '${s.away}-${s.home}').join(', ');

VolleyballStats calculateVolleyballStats(EventsRepository repo, Sport sport) {
  assert(sport == Sport.volleyballMen || sport == Sport.volleyballWomen);

  final overall = VolleyballTeamStats();
  final home = VolleyballTeamStats();
  final away = VolleyballTeamStats();
  final breakdown = VolleyballSetBreakdown();
  final allForm = <VolleyballFormMatch>[];
  final h2hMap = <String, VolleyballHeadToHead>{};
  final scorerTotals = <String, int>{};
  final scorerMatches = <String, int>{};

  var winStreak = 0, lossStreak = 0, longestWinStreak = 0;
  VolleyballRecordResult? biggestWin, heaviestDefeat;
  var biggestWinMargin = 0, biggestWinRally = 0;
  var heaviestDefeatMargin = 0, heaviestDefeatRally = 0;

  for (final month in monthOrder) {
    for (final ev in repo.eventsFor(month)) {
      if (ev.sport != sport || !ev.isPlayed || ev.score == null) continue;
      final parsed = parseScore(ev.score, ev.location);
      if (parsed == null) continue;
      final (setsFor, setsAgainst) = parsed;
      final result = setsFor > setsAgainst ? 'W' : 'L';

      var rallyFor = 0, rallyAgainst = 0;
      for (final set in ev.sets ?? const <VolleyballSet>[]) {
        if (ev.location == MatchLocation.home) {
          rallyFor += set.home;
          rallyAgainst += set.away;
        } else {
          rallyFor += set.away;
          rallyAgainst += set.home;
        }
      }

      for (final stats in [overall, ev.location == MatchLocation.home ? home : away]) {
        stats.played++;
        if (result == 'W') {
          stats.wins++;
        } else {
          stats.losses++;
        }
        stats.setsWon += setsFor;
        stats.setsLost += setsAgainst;
        stats.pointsScored += rallyFor;
        stats.pointsConceded += rallyAgainst;
      }

      switch ('$setsFor-$setsAgainst') {
        case '3-0': breakdown.threeZero++;
        case '3-1': breakdown.threeOne++;
        case '3-2': breakdown.threeTwo++;
        case '0-3': breakdown.zeroThree++;
        case '1-3': breakdown.oneThree++;
        case '2-3': breakdown.twoThree++;
      }

      allForm.add(VolleyballFormMatch(result: result, opponent: ev.opponent, score: ev.score!, location: ev.location, month: month, day: ev.day));

      final h2h = h2hMap.putIfAbsent(ev.opponent, () => VolleyballHeadToHead(ev.opponent));
      h2h.played++;
      if (result == 'W') {
        h2h.wins++;
      } else {
        h2h.losses++;
      }
      h2h.setsWon += setsFor;
      h2h.setsLost += setsAgainst;

      if (result == 'W') {
        winStreak++;
        lossStreak = 0;
        if (winStreak > longestWinStreak) longestWinStreak = winStreak;
      } else {
        lossStreak++;
        winStreak = 0;
      }

      final setMargin = setsFor - setsAgainst;
      final rallyDiff = rallyFor - rallyAgainst;
      final setScores = ev.sets != null ? _formatSetScores(ev.sets!, ev.location) : '';
      if (setMargin > 0 &&
          (biggestWin == null || setMargin > biggestWinMargin || (setMargin == biggestWinMargin && rallyDiff > biggestWinRally))) {
        biggestWin = VolleyballRecordResult(opponent: ev.opponent, score: ev.score!, setScores: setScores);
        biggestWinMargin = setMargin;
        biggestWinRally = rallyDiff;
      }
      if (setMargin < 0 &&
          (heaviestDefeat == null ||
              setMargin.abs() > heaviestDefeatMargin ||
              (setMargin.abs() == heaviestDefeatMargin && rallyDiff < -heaviestDefeatRally))) {
        heaviestDefeat = VolleyballRecordResult(opponent: ev.opponent, score: ev.score!, setScores: setScores);
        heaviestDefeatMargin = setMargin.abs();
        heaviestDefeatRally = rallyDiff.abs();
      }

      // Top scorers: only our own team's players.
      final ourTeam = ev.location == MatchLocation.home ? 'home' : 'away';
      final seenThisMatch = <String>{};
      for (final scorer in ev.vbScorers ?? const <VolleyballScorer>[]) {
        if (scorer.team != ourTeam) continue;
        scorerTotals[scorer.name] = (scorerTotals[scorer.name] ?? 0) + scorer.points;
        seenThisMatch.add(scorer.name);
      }
      for (final name in seenThisMatch) {
        scorerMatches[name] = (scorerMatches[name] ?? 0) + 1;
      }
    }
  }

  final topScorers = scorerTotals.entries
      .map((e) => VolleyballTopScorer(name: e.key, totalPoints: e.value, matchesPlayed: scorerMatches[e.key] ?? 0))
      .toList()
    ..sort((a, b) => b.totalPoints.compareTo(a.totalPoints));

  return VolleyballStats(
    overall: overall,
    home: home,
    away: away,
    setBreakdown: breakdown,
    recentForm: (allForm.length <= 5 ? allForm : allForm.sublist(allForm.length - 5)).reversed.toList(),
    headToHead: h2hMap.values.toList()..sort((a, b) => b.played.compareTo(a.played)),
    currentStreak: winStreak > 0 ? StreakInfo('W', winStreak) : StreakInfo('L', lossStreak),
    longestWinStreak: longestWinStreak,
    biggestWin: biggestWin,
    heaviestDefeat: heaviestDefeat,
    topScorers: topScorers.length > 10 ? topScorers.sublist(0, 10) : topScorers,
  );
}
